Multiple imputation using chained equation (MICE)

Shafaq-Siddiqi opened the pull request:
1. Main DML script (mice_linearReg.dml)
2. Java test file (BuiltinMiceLinearRegTest.java)
3. DML test script (mice_linearRegression.dml)

Shafaq-Siddiqi committed:
Multiple imputation using chained equation (MICE)

1. Main DML script (mice_linearReg.dml)
2. Java test file (BuiltinMiceLinearRegTest.java)
3. DML test script (mice_linearRegression.dml)

phaniarnab commented on scripts/builtin/mice_linearReg.dml:
```diff
@@ -0,0 +1,147 @@
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
```
A modification license is required when modifying an existing file.  Was this algorithm existing before?

phaniarnab commented on scripts/builtin/mice_linearReg.dml:
```diff
@@ -0,0 +1,147 @@
+m_mice_linearReg = function(Matrix[Double] X, Integer iter = 3, Integer complete = 3)
+return(Matrix[Double] dataset, Matrix[Double] singleSet)
+  {
+  n = nrow(X)
```
Please fix the indentation of the opening bracket.

phaniarnab commented on scripts/builtin/mice_linearReg.dml:
```diff
@@ -0,0 +1,147 @@
+  Mask_Filled = Mask;
+
+# filling the missing data with thier means
+  X2 = X+(Mask*mu)
```
You can replace mu with colMeans(X) directly and remove a few unnecessary line spaces throughout.
Please fix the typo in the comment.

phaniarnab commented on scripts/builtin/mice_linearReg.dml:
```diff
@@ -0,0 +1,147 @@
+	  if(i == 1)
+	  {
+	    train_X = slice_X[, (i+1): col]
+	  }
```
You may skip the braces for a single line of codes.

phaniarnab commented on scripts/builtin/mice_linearReg.dml:
```diff
@@ -0,0 +1,147 @@
+	  
+# slicing dataset for filling missing values
+	  slice_Y = removeEmpty(target=X2, margin = "rows", select = Mask[,i])
+	  test_Y = slice_Y[,i]
```
It's a good idea to indent the comment to the same level of the block of code on which it is applied.

phaniarnab commented on src/test/java/org/tugraz/sysds/test/functions/builtin/BuiltinMiceLinearRegTest.java:
```diff
@@ -0,0 +1,85 @@
+public class BuiltinMiceLinearRegTest extends AutomatedTestBase {
+    private final static String TEST_NAME = "mice_linearRegression";
+    private final static String TEST_DIR = "functions/builtin/";
+    private static final String TEST_CLASS_DIR = TEST_DIR + BuiltinMiceLinearRegTest.class.getSimpleName() + "/";
```
Please use tabs for indentation for java files.

phaniarnab commented on src/test/java/org/tugraz/sysds/test/functions/builtin/BuiltinMiceLinearRegTest.java:
```diff
@@ -0,0 +1,85 @@
+import org.tugraz.sysds.test.TestConfiguration;
+import org.tugraz.sysds.test.TestUtils;
+
+import java.util.HashMap;
```
Please remove the unused imports and resolve the warnings.

phaniarnab commented on src/test/scripts/installDependencies.R:
```diff
@@ -35,4 +35,5 @@ custom_install("outliers");
 custom_install("caret");
 custom_install("Sigmoid");
 custom_install("DescTools");
+custom_install("mice");
```
Why is this required?

Shafaq-Siddiqi replied:
Thank you for your comments. I have fixed the issue as per your suggestions. I have also tried to fix variable names.

Shafaq-Siddiqi committed:
MICE Cleanup

1. In java test file spaces are replaced with tabs
2. Indentation in mice_linearReg.dml fixed
3. dependency removed from installDependencies.R

Shafaq-Siddiqi commented:
Values are imputed back using element-wise operations.

mboehm7 commented:
LGTM - thanks @Shafaq-Siddiqi. I further tried to improve the code a bit via vectorized train/test slicing, vectorized imputation, and a parfor loop over columns as these lm models can be trained independently. Furthermore, the imputation seemed to reuse Mask_Filled across iterations such that later iterations had no effect. 

Now I can also reproduce the issues you mentioned (on spark ops, and without the cut between remove empty operations) and will fix them in subsequent commits.
